GuidesTours: Add tours to all signup/new site flows

ehg opened the pull request:
Implements #4378

We want to enable Guided Tours for a subset of new people, across all flows.

This involves modifying the `signup` framework slightly, so we can filter final destinations specified in the flow. It also involves making sure we don't show a tour during checkout.

TODO:
- [x] AB tests: decide on start dates/percentages
- [x] AB test across all flows
- [x] Test all signup flows to make sure they still work, and have decent UX
- [x] Test in desktop app?

To test:
- In console `localStorage.setItem( 'ABTests', '{"guidedTours_20160418":"guided"}' );`
- Add new WordPress when logged in, signup logged out
- Add a plan or theme for some tests
- Check that the tour is shown after signup
- If you've bought something, make sure the tour doesn't show in the checkout, and make sure it appears once you've completed your purchase.

There's also a Calypso-only variant, where we never redirect to the frontend. Test this via:
- `localStorage.setItem( 'ABTests', '{"guidedTours_20160418":"calypsoOnly"}' );`
- Make sure when testing, you land on Calypso, and not the frontend

ehg commented:
@lsinger any ideas about this?

> Decide how we're going to test. Are we going to link to the flow from somewhere? Are we just going to A/B test a percentage of all signups? Do we even need our own flow — we could probably get away with pure A/B testing, as the tour qarg is always read first atm.

lsinger commented:
Is there anything that our own flow would help us with that isn't possible with Calypso's A/B testing?

ehg commented:
> Is there anything that our own flow would help us with that isn't possible with Calypso's A/B testing?

I suppose it could help isolate our tests more from other flows. I guess we should ask ourselves "what do we want to test?" :)

mcsf commented on client/signup/config/flows.js:
```diff
@@ -269,8 +269,16 @@ function filterFlowName( flowName ) {
+	if ( dependenciesContainCartItem( dependencies ) ) {
+		return getCheckoutUrl( dependencies );
+	}
+	return abtest( 'guidedTours' ) ? `/posts/${ dependencies.siteSlug }/?tour=main` : destination;
```
`=== 'guided'`

ehg commented:
@mcsf I implemented a 'delay' for checkout here 105ac91 — sanity check appreciated :)

ehg commented on client/signup/config/flows.js:
```diff
@@ -269,8 +269,23 @@ function filterFlowName( flowName ) {
+			return getCheckoutUrl( dependencies ) + `?tour=${ tourName }`;
+		}
+
+		return `/posts/${ dependencies.siteSlug }/?tour=${ tourName }`;
```
Maybe we can check `isExternal( destination )` here, so for everything but external destinations, we'd land on the original Calypso destination?

lsinger replied:
Sounds like a reasonable guard to me.

ehg commented:
Rebased and decided to allow pre-existing users in the AB-test.

Would appreciate some code review & testing :)

lsinger commented:
Tried to add a new WordPress being logged in. Got stuck on the "Almost done!" screen with the console showing the following:

<img width="679" alt="screen shot 2016-04-12 at 12 30 00 pm" src="https://cloud.githubusercontent.com/assets/23619/14457383/5c209ed4-00aa-11e6-86f2-ed53f07b84c6.png">

I executed the `localStorage` commands before hitting "Add a new WordPress". Anythig I'm doing wrong?

ehg commented:
> Tried to add a new WordPress being logged in. Got stuck on the "Almost done!" screen with the console showing the following[…]

Ooops, sorry. 7273749d11bf3191ad2f7da198836faac322accb should fix it.

lsinger commented:
I went through a few variations:

- tour shows up properly when creating a new site when logged it, even though the "My Site(s)" is visible when we get into Calypso / when the tour starts, which is a bit unfortunate (because nothing changes, really, when you click the first action step)
- as a logged out user, when I set the abtest in localStorage on the landing page, it's gone by the time I've signed up and I get Trampoline
- this also happens if I set the cookie at the last moment before choosing a username and password -- it's there before I click create my account, but then I land on Trampoline and the abtest is gone (also doesn't come back when switching to Calypso); this doesn't seem right
- after purchasing a plan, on the thank you page the most glaring CTA leads to the customizer, which of course doesn't show the tour; when I click "back to my site" instead, no tour is shown even though the abtest is present in localStorage

ehg commented:
Thanks!

> as a logged out user, when I set the abtest in localStorage on the landing page, it's gone by the time I've signed up and I get Trampoline

Have you tried setting the abtest once you're in Calypso signup? (e.g. /start). Also, which landing page are you using?

lsinger commented:
> Have you tried setting the abtest once you're in Calypso signup? (e.g. /start). Also, which landing page are you using?

Yup, see the bullet point after that one. I used the blue default one (i.e. not delta-new).

mcsf commented on client/signup/config/flows.js:
```diff
@@ -269,8 +270,29 @@ function filterFlowName( flowName ) {
 
+function filterDestination( destination, dependencies, flowName ) {
+	if ( abtest( 'guidedTours' ) === 'guided' ) {
+		const disabledForFlows = [ 'account', 'site-user' ];
```
`disabledFlows`?

ehg replied:
6722db8

mcsf commented on client/signup/config/flows.js:
```diff
@@ -269,8 +270,29 @@ function filterFlowName( flowName ) {
+
+		const tourName = 'main';
+		//TODO: Build the query arg properly, so we don't have problems with
+		// destinations that already have query strings in
```
Maybe sign this TODO, so we can get pinged if we ever forget? (Also, spacing before "TODO", and place it right before the return statement rather than after the `tourName` declaration).

ehg replied:
6722db8

mcsf commented on client/signup/main.jsx:
```diff
@@ -96,12 +96,11 @@ const Signup = React.createClass( {
 				}
-
-				return this.handleFlowComplete( dependencies, destination );
+				return this.handleFlowComplete( dependencies, utils.getDestination( destination, dependencies, this.props.flowName ) );
```
Shame we need to manually glue `getDestination` to each occurrence of `handleFlowComplete`. Could we compose higher up or something?

ehg replied:
581730a look OK?

mcsf replied:
ta!

mcsf commented:
Code looks good. However, one question+suggestion: is `shouldShow`, as exposed by our selector, ever consumed by anything? If not, why expose it? My thinking is inspired in this bit from another PR:

```js
export function getGuidesTourState( state ) {
	const { shouldShow, stepName = '' } = state.ui.guidesTour;
	const stepConfig = guidesToursConfig[ stepName ] || false;
	const shouldReallyShow = shouldShow && !! getSelectedSiteId( state );
	return Object.assign( {}, state.ui.guidesTour, {
		stepConfig,
		shouldShow: shouldReallyShow,
		showPreview: shouldReallyShow && stepConfig.showPreview,
	} );
}
```

Note how, in the returned object, we override `state.ui.guidesTour`'s own `shouldShow` with our own version (`shouldReallyShow`, which considers both `shouldShow` and external data).

That way, instead of forcing a developer to know the difference between `shouldShow` and `willShow`, the selector calls the shots and only exposes its interpretation of `shouldShow`.

Would that make sense?

lsinger commented:
> Also, which landing page are you using? 

I'm an idiot. I went to WordPress.com to sign up. Using local Calypso the flow works. 😡

lsinger commented:
From my comments these I think are still valid, but probably belong into a different PR and need separate discussions: 

- The sidebar is visible when we get into Calypso / when the tour starts. This is unfortunate because nothing changes, really, when you click the first action step. If the first step needs to introduce the sidebar AND it is visible, it probably shouldn't be an action step. /cc @shaunandrews 
- On the thank you page the most glaring CTA leads to the customizer, which of course doesn't show the tour. Either ignore (they'll get into Calypso in time) or talk to whoever is working on the thankyou page.

ehg commented:
> On the thank you page the most glaring CTA leads to the customizer, which of course doesn't show the tour. Either ignore (they'll get into Calypso in time) or talk to whoever is working on the thankyou page.

Good spot. Interestingly, the tour will probably show in the Customizer, as it's just an overlay iframe. It'll look like this, though…

![screen shot 2016-04-13 at 21 05 59](https://cloud.githubusercontent.com/assets/215074/14507676/b20ea40a-01bb-11e6-9cfb-22ecb146e009.png)

Maybe we could delay the tour ( have the steps shoot off the screen ), until the Customizer is exited?

shaunandrews commented:
I've updated the prototype to account for the lack of a preview. Users would start on My Site with the sidebar already in view: https://marvelapp.com/i9414d

That means no need for the "click here" step.

ehg commented:
> That way, instead of forcing a developer to know the difference between shouldShow and willShow, the selector calls the shots and only exposes its interpretation of shouldShow.

Rebased, and added a5f2671 — does it make sense?

lsinger commented on client/signup/config/flows.js:
```diff
@@ -269,8 +270,28 @@ function filterFlowName( flowName ) {
+		// TODO(ehg): Build the query arg properly, so we don't have problems with
+		// destinations that already have query strings in
+		return isOutsideCalypso( destination )
+			? `/stats/${ dependencies.siteSlug }?tour=${ tourName }`
```
Naive question: is `dependencies.siteSlug` guaranteed to be set here?

ehg replied:
Probably, but I put a guard in a8798e2 :)

mcsf commented:
> Rebased, and added a5f2671 — does it make sense?

It does, thanks.

ehg commented:
Another rebase, and added a8798e2

mcsf commented on client/layout/index.jsx:
```diff
@@ -184,13 +185,13 @@ Layout = React.createClass( {
 			section,
 			isOffline: isOffline( state ),
-			shouldShowGuidesTour: guidesTour.shouldShow,
+			guidesState: getGuidesTourState( state ),
```
[borderline nitpicky] Other GT-connected pieces use the name `tourState`. Not saying we need to change it to that, but we should be aware that we have different names for the same data lying around.

mcsf commented:
I'm at f673fa and I'm seeing the tour cut short unexpectedly:

![calypso-guidestours-160415](https://cloud.githubusercontent.com/assets/150562/14560732/71e8c10e-0309-11e6-9c3f-4a01ba0a8da4.gif)

mcsf commented:
LGTM. Squash and merge?

ehg commented:
So the problem with merging this, is that it'll start the abtest straight away. I wonder if we need a disabled flag in `active-tests`. :)

ehg commented:
> So the problem with merging this, is that it'll start the abtest straight away.

Added this: a94374f — make sense?

lsinger commented:
@ehg is that datestamp definitely respected, or is it just for documentation? Alternatively, would setting the guides variation to `0` work?

ehg commented:
> is that datestamp definitely respected, or is it just for documentation? Alternatively, would setting the guides variation to 0 work?

https://github.com/Automattic/wp-calypso/blob/master/client/lib/abtest/index.js#L75

ehg commented:
Looks like we can't both disable the abtest, and test it, so I've set the date to today. I guess we can update the date when we ship to production, so we get a clear demarcation for data.

mcsf commented:
> Looks like we can't both disable the abtest, and test it, so I've set the date to today. I guess we can update the date when we ship to production, so we get a clear demarcation for data.

Sounds good.

ehg commented:
Confirmed it doesn't break desktop in scenarios:
- abtest not active
- abtest active/guidestours flag disabled
- abtest active/guidestours flag enabled - (it works!)
